File: Hangman.js
const words = ["python", "hangman", "programming", "computer", "algorithm", "developer", "challenge"];
let word = "";
let wordLetters = new Set();
let usedLetters = new Set();
let tries = 6;

const hangmanParts = ["head", "body", "left-arm", "right-arm", "left-leg", "right-leg"];

function chooseWord() {
    word = words[Math.floor(Math.random() * words.length)];
    wordLetters = new Set(word);
}

function updateDisplay() {
    const wordDisplay = document.getElementById('word-display');
    wordDisplay.innerHTML = word.split('').map(letter => 
        `<span class="letter ${usedLetters.has(letter) ? 'revealed' : ''}">${usedLetters.has(letter) ? letter : '_'}</span>`
    ).join(' ');

    document.getElementById('used-letters').textContent = 'Used letters: ' + Array.from(usedLetters).join(', ');
}

function updateHangman() {
    const partsToShow = 6 - tries;
    hangmanParts.forEach((part, index) => {
        document.getElementById(part).style.opacity = index < partsToShow ? 1 : 0;
    });
}

function guessLetter() {
    const guessInput = document.getElementById('guess-input');
    const letter = guessInput.value.toLowerCase();
    guessInput.value = '';

    if (letter.length !== 1 || !/[a-z]/.test(letter)) {
        showMessage('Please enter a valid letter.', 'warning');
        return;
    }

    if (usedLetters.has(letter)) {
        showMessage('You have already used that letter.', 'warning');
        return;
    }

    usedLetters.add(letter);

    if (wordLetters.has(letter)) {
        wordLetters.delete(letter);
        showMessage('Good guess!', 'correct');
    } else {
        tries--;
        updateHangman();
        showMessage(`Wrong guess! ${tries} tries left.`, 'wrong');
    }

    updateDisplay();

    if (tries === 0) {
        endGame(false);
    } else if (wordLetters.size === 0) {
        endGame(true);
    }
}

function showMessage(text, className) {
    const messageEl = document.getElementById('message');
    messageEl.textContent = text;
    messageEl.className = className;
}

function endGame(isWin) {
    const message = isWin ? 'Congratulations! You won!' : `Game over! The word was "${word}".`;
    showMessage(message, isWin ? 'win' : 'game-over');
    document.getElementById('guess-button').disabled = true;
    document.getElementById('guess-input').disabled = true;
    document.getElementById('restart-button').style.display = 'inline-block';
}

function restartGame() {
    tries = 6;
    usedLetters.clear();
    chooseWord();
    updateDisplay();
    updateHangman();
    showMessage('', '');
    document.getElementById('guess-button').disabled = false;
    document.getElementById('guess-input').disabled = false;
    document.getElementById('restart-button').style.display = 'none';
}

function initGame() {
    chooseWord();
    updateDisplay();
    updateHangman();
    document.getElementById('guess-button').addEventListener('click', guessLetter);
    document.getElementById('guess-input').addEventListener('keyup', function(event) {
        if (event.key === 'Enter') {
            guessLetter();
        }
    });
    document.getElementById('restart-button').addEventListener('click', restartGame);
}

initGame();

File: Hangman.py
import random

def choose_word():
    words = ["python", "hangman", "programming", "computer", "algorithm", "developer", "challenge"]
    return random.choice(words)

def display_hangman(tries):
    stages = [  # final state: head, torso, both arms, and both legs
                """
                   --------
                   |      |
                   |      O
                   |     \\|/
                   |      |
                   |     / \\
                   -
                """,
                # head, torso, both arms, and one leg
                """
                   --------
                   |      |
                   |      O
                   |     \\|/
                   |      |
                   |     / 
                   -
                """,
                # head, torso, and both arms
                """
                   --------
                   |      |
                   |      O
                   |     \\|/
                   |      |
                   |      
                   -
                """,
                # head, torso, and one arm
                """
                   --------
                   |      |
                   |      O
                   |     \\|
                   |      |
                   |     
                   -
                """,
                # head and torso
                """
                   --------
                   |      |
                   |      O
                   |      |
                   |      |
                   |     
                   -
                """,
                # head
                """
                   --------
                   |      |
                   |      O
                   |    
                   |      
                   |     
                   -
                """,
                # initial empty state
                """
                   --------
                   |      |
                   |      
                   |    
                   |      
                   |     
                   -
                """
    ]
    return stages[tries]

def hangman():
    word = choose_word()
    word_letters = set(word)  # letters in the word
    alphabet = set("abcdefghijklmnopqrstuvwxyz")
    used_letters = set()  # what the user has guessed
    tries = 6

    print("Welcome to Hangman!")
    
    while len(word_letters) > 0 and tries > 0:
        # letters used
        print("You have used these letters: ", ' '.join(used_letters))

        # current word
        word_list = [letter if letter in used_letters else '_' for letter in word]
        print("Current word: ", ' '.join(word_list))

        user_letter = input("Guess a letter: ").lower()

        if user_letter in alphabet - used_letters:
            used_letters.add(user_letter)
            if user_letter in word_letters:
                word_letters.remove(user_letter)
            else:
                tries -= 1
                print(f"Letter {user_letter} is not in the word.")
                print(display_hangman(tries))

        elif user_letter in used_letters:
            print("You have already used that letter. Guess another letter.")

        else:
            print("Invalid character. Please enter a letter.")

    if tries == 0:
        print(f"Sorry, you lost. The word was {word}.")
    else:
        print(f"Congratulations! You guessed the word {word}!")

if __name__ == "__main__":
    hangman()
